labgrid/util/snmp.py

```diff
@@ -8,24 +10,39 @@ def __init__(self, host, community, port=161):
         if port is None:
             port = 161
 
+        try:
+            # if called from async code, try to get current's thread loop
+            self.loop = asyncio.get_running_loop()
+        except RuntimeError:
+            # no previous, external or running loop found, create a new one
+            self.loop = asyncio.new_event_loop()
```

Comment:
Client code does the same, but only in `start_session()` when no loop is provided otherwise (and stashes the loop for later use).

How would other code retrieve the newly created event loop from here?

I think this must rely on an existing event loop, similar to EthernetPortManager:

https://github.com/labgrid-project/labgrid/blob/6b541210c6063e97d81e257d2d6bda1444d9ec78/labgrid/resource/ethernetport.py#L251

Reply:
But `get_event_loop` is deprecated, so I chose not to use this.

Reply:
No, `get_event_loop()` is not deprecated per se: [only if there is no current event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop). That's why I think we should simply expect a loop here. If there is none, that's an error that needs to be fixed outside of labgrid.

Reply:
Using labgrid as a library means that there is not necessarily an available loop and we need to support this (which is exactly the case where there is no loop).

Reply:
Testing this inside of labgrid client reveals that `get_running_loop` does not find the running client event loop and instead no loop is found and another is created for SNMP.

Reply:
> Using labgrid as a library means that there is not necessarily an available loop and we need to support this (which is exactly the case where there is no loop).

I disagree. There are certain "entry points" into labgrid where a loop can be passed or labgrid tries to find one or sets a new one up as a fallback. Exporter, coordinator and client are these "entry points". Client and libraries typically use `labgrid.remote.client.start_session()` to start a session.  So that makes sure there is a loop.

For all other cases, code that uses labgrid as a library needs to make sure a loop exists. Maybe that needs to be documented better.

Using `get_event_loop()` in EthernetPortManager was an active decision. Using that makes sure there is an event loop. Otherwise you get a warning (and a new event loop) or an error, depending on the Python version.

> Testing this inside of labgrid client reveals that `get_running_loop` does not find the running client event loop and instead no loop is found and another is created for SNMP.

That's because `get_running_loop()` does not work in sync contexts and there is no other way to achieve that.

Reply:
I guess this is fine for the SNMP switch handling, but isn't the correct requirement for the simple power backend handling we had beforehand. So how do we proceed with this? Requiring an event loop for sync library usage is not really feasible.

Reply:
How does the simple power backend handling work without a session?

Reply:
> How does the simple power backend handling work without a session?

What do you mean? You can totally describe all of the `NetworkPowerPort`s inside of the YAML configuration and not use a `RemotePlace` to retrieve the resources.

Reply:
Yes, I see.

See commit messages of da95c444, 0e1e4fba and 606af353 for context.

What about calling `labgrid.remote.client.ensure_event_loop()` whenever we need an event loop? That would restore the original behavior of using labgrid as a library without a coordinator session.

We could add a function that allows library users to pass their event loop into labgrid (stashing it) for use cases without `labgrid.remote.client.start_session()` (which already allows passing an external loop).